ethereum/trie/src/test/java/tech/pegasys/pantheon/ethereum/trie/TrieNodeDecoderTest.java

```diff
@@ -0,0 +1,224 @@
+    assertThat(actualValues)
+        .containsExactly(
+            BytesValue.of(1),
+            BytesValue.of(10),
+            BytesValue.of(2),
+            BytesValue.of(20),
+            BytesValue.of(3),
+            BytesValue.of(30));
+  }
+
+  @Test
+  public void breadthFirstDecode_partialTrie() {
+    final InMemoryKeyValueStorage fullStorage = new InMemoryKeyValueStorage();
+    final InMemoryKeyValueStorage partialStorage = new InMemoryKeyValueStorage();
+
+    // Build a small trie
+    MerklePatriciaTrie<BytesValue, BytesValue> trie =
+        new StoredMerklePatriciaTrie<>(fullStorage::get, Function.identity(), Function.identity());
+    final Random random = new Random(1);
+    for (int i = 0; i < 30; i++) {
+      byte[] key = new byte[4];
+      byte[] val = new byte[4];
+      random.nextBytes(key);
+      random.nextBytes(val);
+      trie.put(BytesValue.wrap(key), BytesValue.wrap(val));
+    }
+    final Transaction tx = fullStorage.startTransaction();
+    trie.commit(tx::put);
+    tx.commit();
+
+    // Get root node
+    Node<BytesValue> rootNode =
+        TrieNodeDecoder.breadthFirstDecoder(fullStorage::get, trie.getRootHash()).findFirst().get();
+
+    // Decode partially available trie
+    final Transaction partialTx = partialStorage.startTransaction();
+    partialTx.put(trie.getRootHash(), rootNode.getRlp());
+    partialTx.commit();
+    final List<Node<BytesValue>> allDecodableNodes =
+        TrieNodeDecoder.breadthFirstDecoder(partialStorage::get, trie.getRootHash())
+            .collect(Collectors.toList());
+    assertThat(allDecodableNodes.size()).isGreaterThanOrEqualTo(1);
+    assertThat(allDecodableNodes.get(0).getHash()).isEqualTo(rootNode.getHash());
+  }
+
+  @Test
+  public void breadthFirstDecode_emptyTrie() {
+    List<Node<BytesValue>> result =
+        TrieNodeDecoder.breadthFirstDecoder(
+                (h) -> Optional.empty(), MerklePatriciaTrie.EMPTY_TRIE_NODE_HASH)
+            .collect(Collectors.toList());
+    assertThat(result.size()).isEqualTo(0);
+  }
+
+  @Test
+  public void breadthFirstDecode_singleNodeTrie() {
+    final InMemoryKeyValueStorage storage = new InMemoryKeyValueStorage();
+
+    MerklePatriciaTrie<BytesValue, BytesValue> trie =
+        new StoredMerklePatriciaTrie<>(storage::get, Function.identity(), Function.identity());
+    trie.put(BytesValue.fromHexString("0x100000"), BytesValue.of(1));
+
+    // Save nodes to storage
+    final Transaction tx = storage.startTransaction();
+    trie.commit(tx::put);
+    tx.commit();
+
+    List<Node<BytesValue>> result =
+        TrieNodeDecoder.breadthFirstDecoder(storage::get, trie.getRootHash())
+            .collect(Collectors.toList());
+    assertThat(result.size()).isEqualTo(1);
+    assertThat(result.get(0).getValue()).contains(BytesValue.of(1));
+    // TODO: fix path representation
+    // assertThat(result.get(0).getPath()).isEqualTo(BytesValue.fromHexString("0x100000"));
```

Comment:
Does this need looking at still?

Reply:
Yeah - I'll make a ticket for this.  It's an existing issue.

Reply:
Looks like this isn't actually a "bug" - the path value is just handled in an idiosyncratic way.  Updated the test.